desc/objectives/_generic.py

```diff
@@ -15,6 +19,262 @@
 from .objective_funs import _Objective
 
 
+class _ExternalObjective(_Objective, ABC):
+    """Wrap an external code.
+
+    Similar to ``ObjectiveFromUser``, except derivatives of the objective function are
+    computed with finite differences instead of AD.
+
+    The user supplied function must take an Equilibrium as its only positional argument,
+    but can take additional keyword arguments.
+
+    # TODO: add Parameters documentation
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    fun : callable
+        Custom objective function.
+    dim_f : int
+        Dimension of the output of fun.
+    target : {float, ndarray}, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        Must be broadcastable to Objective.dim_f. Defaults to ``target=0``.
+    bounds : tuple of {float, ndarray}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to to Objective.dim_f.
+        Defaults to ``target=0``.
+    weight : {float, ndarray}, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+        Has no effect for this objective.
+    normalize_target : bool, optional
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization.
+    name : str, optional
+        Name of the objective function.
+
+    # TODO: add example
+
+    """
+
+    _units = "(Unknown)"
+    _print_value_fmt = "External objective value: {:10.3e}"
+
+    def __init__(
+        self,
+        eq,
+        fun,
+        dim_f,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=False,
+        normalize_target=False,
+        loss_function=None,
+        fd_step=1e-4,  # TODO: generalize this to allow a vector of different scales
```

Comment:
maybe just have an absolute and relative step size? so that `dx = abs_step + rel_step*abs(x)`

not sure we want to deal with different step sizes for each element in x and figuring out how they get mapped after projections etc.

Reply:
actually I don't think a different fd step per element would even work with JVPs, since using different steps for different components effectively changes the direction of the JVP.

Reply:
@f0uriest Do you mean: `dx = abs_step + rel_step * np.mean(np.abs(x))` or `dx = abs_step + rel_step * np.abs(x)`? The former is a scalar and the latter is a vector. 

Also would this be in addition to your other comment of scaling `self._fd_step` by `norm(x)` within the `_jaxify` function, or would this be instead of that?